src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+
+use mz_repr::adt::datetime::DateTimeUnits;
+use mz_repr::{ColumnType, Datum, RelationType, Row, RowArena, ScalarType};
+
+use crate::{
+    BinaryFunc, EvalError, MapFilterProject, MfpPlan, MirScalarExpr, UnaryFunc,
+    UnmaterializableFunc, VariadicFunc,
+};
+
+/// Tracks whether a function is monotonic, either increasing or decreasing.
+/// This property is useful for static analysis because a monotonic function maps ranges to ranges:
+/// ie. if `a` is between `b` and `c` inclusive, then `f(a)` is between `f(b)` and `f(c)` inclusive.
+/// (However, if either `f(b)` and `f(c)` error or return null, we don't assume anything about `f(a)`.
+///
+/// This should likely be moved to a function annotation in the future.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum Monotonic {
+    /// Monotonic in this argument.
+    Yes,
+    /// We haven't classified this one yet. (Conservatively treated as "no", but a distinct
+    /// enum so it's easy to tell whether someone's thought about a particular function yet.)
+    /// This should likely be removed when this becomes a function annotation.
+    Maybe,
+    /// We don't know any useful properties of this function.
+    No,
+}
+
+fn unary_monotonic(func: &UnaryFunc) -> Monotonic {
+    use crate::func::impls as funcs;
+    use Monotonic::*;
+    use UnaryFunc::*;
+    match func {
+        // Casts are generally monotonic.
+        CastUint64ToNumeric(_)
+        | CastNumericToUint64(_)
+        | CastNumericToMzTimestamp(_)
+        | CastUint64ToMzTimestamp(_)
+        | CastTimestampToMzTimestamp(_)
+        | CastTimestampTzToMzTimestamp(_) => Yes,
+        // Including JSON casts.
+        CastJsonbToNumeric(_)
+        | CastJsonbToBool(_)
+        | CastJsonbToString(_)
+        | CastJsonbToInt64(_)
+        | CastJsonbToFloat64(_) => Yes,
+
+        // Extracting the "most significant bits" of the a timestamp is monotonic.
+        ExtractTimestamp(funcs::ExtractTimestamp(units))
+        | ExtractTimestampTz(funcs::ExtractTimestampTz(units)) => match units {
+            DateTimeUnits::Epoch
+            | DateTimeUnits::Millennium
+            | DateTimeUnits::Century
+            | DateTimeUnits::Decade
+            | DateTimeUnits::Year => Yes,
+            _ => No,
+        },
+        DateTruncTimestamp(_) | DateTruncTimestampTz(_) => Yes,
+        // Negation is monotonically decreasing, but that's fine.
+        NegInt64(_) | NegNumeric(_) | Not(_) => Yes,
+        // Trivially monotonic, since all non-null values map to `false`.
+        IsNull(_) => Yes,
+        AbsInt64(_) => No,
+        _ => Maybe,
```

Comment:
While it's more verbose, writing out every function will make sure that any new functions get consideration here.